### PR TITLE
Implement Vendor extensions.

### PR DESCRIPTION
**Pull request by glg-rv**
This patch allows SBI users to create vendor extensions to wrap the standard SBI implementation.

An `enum` `T` implementing trait `VendorExtension` will be able to be be used in `VendorSbiMessage<T>`, and be parsed as SBI messages in code using this library.

As an example, this [change] would allow `salus` to use `SalusSbiMessage` (alias for `VendorSbiMessage<SalusExtension`) to parse salus-specific vendor extensions.

This is a RFC.  

[change]: https://github.com/glg-rv/salus/commit/34d42d908520c6be556bd8127ae1eb737d8f60c3

### CODE REVIEW

File: src/salus.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    fn a5(&self) -> u64 {
+        use SalusExtension::*;
+        match self {
+            SalusTest(m) => m.a5(),
+        }
+    }
+
+    fn a6(&self) -> u64 {
```
Comment (tdewey-rivos) [inline src/salus.rs:79]: Increasingly keen on turning these into a macro!.

File: src/vendor.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2023 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::consts::*;
+use crate::error::*;
+use crate::function::*;
+use crate::SbiMessage;
+
+pub trait VendorExtension: SbiFunction + Sized {
```
Comment (tdewey-rivos) [inline src/vendor.rs:10]: Very sensible 👍

**Comment by tdewey-rivos** [comment]
I like it - it gets closer to what I was aiming for, and still gives the full passthrough you'd want for Vendor calls. Assuming we're happy with this pattern, I'd also be keen to see a PR for the Firmware-specific range and Experimental range.

File: src/salus.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        match self {
+            SalusTest(m) => m.a6(),
+        }
+    }
+}
+
+/// Functions defined for the Rivos test extension
+#[derive(Clone, Copy, Debug)]
+pub enum SalusTestFunction {
+    /// Returns the implemented version of the SBI standard.
```
Comment (dgreid) [inline src/salus.rs:90]: stale comment?

File: src/salus.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+use crate::error::*;
+use crate::vendor::*;
+use crate::SbiFunction;
+
+const EXT_SALUS_TEST: u64 = 0x09FFFFFF;
+
+pub type SalusSbiMessage = VendorSbiMessage<SalusExtension>;
+
+pub enum SalusExtension {
```
Comment (dgreid) [inline src/salus.rs:15]: Why is SalusExtension an enum? This extension is for a7 == 0x09FFFFFF (Test), any other extensions added will need a different a7 and be a different extension, just like another vendors specific extension. Or am I missing something?

**Review by dgreid — COMMENTED** [review]
Thanks, I like this approach. Just confused on how extensions will be grouped.

File: src/salus.rs (hunk 4 above)
Comment (dgreid) [reply src/salus.rs:15]: Ah, that makes sense. Maybe "SalusExtenions" plural then?

For Rivos we'll likely have different but overlapping sets of extensions for m-mode FW, s-mode FW, and Salus. How would those be shared? I'm considering the trade off between adding one file per extension and one file per SBI implementation. Thoughts?

Comment (glg-rv) [reply src/salus.rs:15]: So if they're different but overlapping I guess this is kind of what I was going for, at least for now.
Anyone using `sbi-rs` doesn't have to use the salus extensions. Has to import the module manually.
Probably the code will have to be restructured in the future, but for now I was thinking of keeping all salus functions for all salus extensions in that file, and then possibly move them in a module with its own directory (salus/mod.rs) when it grows out of a sane single file source.
There's also the possibility for the user of `sbi-rs` would implement the set of extensions outside of `sbi-rs`, and then how it implements that will be project specific.

Comment (dgreid) [reply src/salus.rs:15]: sgtm. @tdewey-rivos does this work for you? If so, we'll merge this one, nothing set in stone, still at 0.1 so APIs can evolve as needed.

Comment (glg-rv) [reply src/salus.rs:15]: am about to push some changes to this PR allow API calls.

**Comment by glg-rv** [comment]
Pushed new series of patches. Apart from fixing some typos and stale comment, all the changes here are motivated by the need of adding API functions.

A new trait, `EcallMessage` has been added to clarify the interface needed to use `ecall_send`, as simply passing `SbiMessage` was not an option now that we can use `VendorSbiMessage`.

Personally fell into a deep hole when I tried to rationalize the interfaces (`VendorExtensions` and the set of standard SBI extensions are the same interface), and decided to have a simple change first, and simplify (if they are actually simplified) things later.

Same story for removing all the boilerplate code. I experimented with `enum_dispatch`, it does help but has the effect of deciding the field names of the `enum` and makes the code a bit hard to follow. Not a big issue, but decided not to use it in this change to make these changes simple and self-evident. Same discourse with `paste`.

TLDR; removal of boilerplate code is left as a gift to future us.

File: src/salus.rs (hunk 1 above)
Comment (glg-rv) [reply src/salus.rs]: Agreed. This will either require a `enum_dispatch` or `paste` as suggested. I found doing this being a distraction of what this patch should do -- made the code slightly harder to follow, so postponed to future work.

File: src/salus.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Salus vendor exception.
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/salus.rs]: Fixed.

File: src/salus.rs (hunk 3 above)
Comment (glg-rv) [reply src/salus.rs]: Yes. :)

**Review by atulkharerivos — COMMENTED** [review]
Very nice...the only question in my mind is whether we'll need to track allocation of the vendor ranges. Ideally, we would want to try and keep vendor ranges somewhat predictable (a la PCI vendor ID). Perhaps, starting a separate git repository to track vendor IDs might nudge things in the right direction, but it's probably too early for it?

**Comment by glg-rv** [comment]
> Very nice...the only question in my mind is whether we'll need to track allocation of the vendor ranges. Ideally, we would want to try and keep vendor ranges somewhat predictable (a la PCI vendor ID). Perhaps, starting a separate git repository to track vendor IDs might nudge things in the right direction, but it's probably too early for it?

Possibly not too early, but this is comfortably orthogonal to this patch. This just allows to have each project its own vendor range. This patch arises from the need for salus to have its own test functions, so that other users might follow.

**Comment by dgreid** [comment]
I'm not sure about the API changes. Could sbi-rs expose something raw such as a 'Vendor' SbiMessage that is constructed with an extension ID and a `impl SbiFunction`?

Similarly on the decode side a "Vendor" variant with bare registers could be parsed and then the SBI implementor could do vendor-specific match statements. For salus this would be a handle_vendor_msg function added to vm.rs.

**Comment by tdewey-rivos** [comment]
> Very nice...the only question in my mind is whether we'll need to track allocation of the vendor ranges. Ideally, we would want to try and keep vendor ranges somewhat predictable (a la PCI vendor ID). Perhaps, starting a separate git repository to track vendor IDs might nudge things in the right direction, but it's probably too early for it?

My understanding of the wider SBI spec is that the Vendor range is allocated per the low bits of mvendorid on-device, which is in turn allocated by JEDEC:

https://github.com/riscv-non-isa/riscv-sbi-doc/blob/master/riscv-sbi.adoc#vendor-specific-sbi-extension-space-eids-0x09000000-0x09ffffff

https://github.com/riscv/riscv-isa-manual/blob/master/src/machine.tex#L220

**Comment by glg-rv** [comment]
> I'm not sure about the API changes. Could sbi-rs expose something raw such as a 'Vendor' SbiMessage that is constructed with an extension ID and a `impl SbiFunction`?
> 
> Similarly on the decode side a "Vendor" variant with bare registers could be parsed and then the SBI implementor could do vendor-specific match statements. For salus this would be a handle_vendor_msg function added to vm.rs.

Yes! Thank you for this. Navigating through the possibility space of Rust types is still confusing for me. But this is definitely an excellent local minimum. 

I updated the PR with your suggestion, or at least what I think you meant, and it is much simpler. Also leaves total freedom to the user of this crate on how to reconstruct and interpret the message.

**Review by dgreid — APPROVED** [review]
Thanks!

File: src/ecall.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        // Default implementation.
+        SbiReturn {
+            error_code: a0 as i64,
+            return_value: a1,
+        }
+        .into()
+    }
+}
+
+/// Send an ecall to the firmware or hypervisor.
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/ecall.rs]: (super)Fixed. This was actually not on my code, but added to the fix typos commit

**Comment by dgreid** [comment]
I'll merge this tomorrow to give @tdewey-rivos a chance to comment first. Thanks!

File: src/sbi.rs [hunk 7]
```diff
@@ -366,6 +382,11 @@ pub unsafe fn ecall_send(msg: &SbiMessage) -> Result<u64> {
 }
 
 #[cfg(not(all(target_arch = "riscv64", target_os = "none")))]
-unsafe fn ecall_send(_msg: &SbiMessage) -> Result<u64> {
+/// Test Compilation only.
```
Comment (tdewey-rivos) [inline src/sbi.rs:385]: Was there any particular reason we couldn't wrap this in #cfg(test)?

**Comment by tdewey-rivos** [comment]
LGTM. One curiosity, but not enough to block a merge. Thanks for digging in, Gianluca.

File: src/sbi.rs (hunk 7 above)
Comment (glg-rv) [reply src/sbi.rs:385]: Wondered the same, but in general I think this is just a nice _contrapasso_ to the previous CFG for the `ecall_send`. :-)

Comment (dgreid) [reply src/sbi.rs:385]: Ah, this isn't exactly test only. It is linked when I use this crate from a userspace VMM. The VMM can't make tee ecalls, it just makes syscalls. but it does need the rest of the crate, so panic is the right behavior in that case.

**Comment by atishp04** [comment]
@tdewey-rivos raised a valid point about vendor extension IDs. I don't see it being addressed. Please let me know if I missed something.

As per the SBI spec the vendor extension should have lower 24 bits from mvendorid. Thus, every vendor can have only 1 extension ID. We also need to define some kind of sub-extension or fid groups probing mechanism to differentiate between different rivos specific extensions/function groups.

Sorry for posting the comments after it got merged but I missed the notification in gmail as it went to other folders.

**Comment by glg-rv** [comment]
Sure, I'll follow the conversation of FID allocation. Once we get into the point of actually defining what the strategy for ID allocation for project specific testing is, then I'll change this code. As it is now, being used only for testing, is not creating any harm.
